alarm_active_table.cpp

```diff
@@ -0,0 +1,453 @@
+                               context->_alarm_table_def->description().length());
+    }
+    break;
+
+    case COLUMN_ALARMACTIVELOGPOINTER:
+    {
+      // We don't keep log entries so this will always be
+      // the 0.0 OID
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) zero_dot_zero_oid,
+                               sizeof(zero_dot_zero_oid));
+    }
+    break;
+
+    case COLUMN_ALARMACTIVEMODELPOINTER:
+    {
+      oid* model_pointer;
+      // We add 3 here to accomodate for the leading "0" (to be inserted before
+      // the alarm's index), the alarm's index and the alarm's severity.
+      int model_pointer_len = alarmModelTable_oid_len + entry_column_oid_len + 3;
+      // Append the index array on to the Alarm Model Table OID, inserting
+      // the OID ".1.3." for the entry in the first non index column and
+      // inserting a leading "0" before the alarm's index.
+      model_pointer = new oid[model_pointer_len];
+      copy(alarmModelTable_oid, alarmModelTable_oid + alarmModelTable_oid_len, model_pointer);
+      copy(entry_column_oid, entry_column_oid + entry_column_oid_len, model_pointer + alarmModelTable_oid_len);
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len] = 0;
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len + 1] = context->_alarm_table_def->alarm_index(); 
+      model_pointer[alarmModelTable_oid_len + entry_column_oid_len + 2] = context->_alarm_table_def->severity();      
+      
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) model_pointer,
+                               model_pointer_len * sizeof(oid));
+      delete[](model_pointer);
+
+    }
+    break;
+    
+    case COLUMN_ALARMACTIVESPECIFICPOINTER:
+    {
+      // This would point to the model-specific active alarm list
+      // for this alarm. We do not support model-specific Alarm-MIB
+      // and hence this will always be the 0.0 OID
+      snmp_set_var_typed_value(var, ASN_OBJECT_ID,
+                               (u_char*) zero_dot_zero_oid,
+                               sizeof(zero_dot_zero_oid));
+    }
+    break;
+    
+    
+    default: /** We shouldn't get here */
+    {
+      TRC_ERROR("unknown column: alarmActiveTable_get_value");
+      return SNMP_ERR_GENERR;
+    }
+  }
+  return SNMP_ERR_NOERROR;
+}
+
+/*************************************************************
+ *
+ * Find the current time and use it to fill the d_time struct
+ * which is formatted to be compliant with RFC 3877
+ */
+alarmActiveTable_SNMPDateTime alarm_time_issued(void)
+{
+  alarmActiveTable_SNMPDateTime d_time;
+  struct timespec ts;
+  struct tm *timing = new tm;
+  time_t rawtime;
+  int UTC_offset;
+
+  clock_gettime(CLOCK_REALTIME, &ts);
+  rawtime = ts.tv_sec;
+  // localtime_r converts calendar time to a broken down representation
+  // to be stored in the stucture timing
+  localtime_r(&rawtime, timing);
```

Comment:
I don't think you need to call both `localtime` and `localtime_r` - that looks odd. (I see that `localtime_r` "need not set tzname, timezone, and daylight", which you need, so I'd probably just call localtime, with a comment explaining why we can't call `localtime_r`).